example/amagaki.yaml

```diff
@@ -8,6 +8,9 @@ localization:
   - it
   - ja
 
+plugins:
```

Comment:
After having thought about it more the last few days (and I'm not 100% sold on what I'm about to say, I just want to discuss it further), I don't know if the YAML file is the right place for plugin management. I think we might want to avoid the need to enumerate plugins in YAML, and just stick to loading plugins entirely in JS. The YAML thing is another proprietary concept that JS users would have to figure out and learn.

After several years of usage I never liked how in Grow I had to both edit the `podspec.yaml` and also add a bunch of boilerplate stuff in a local `extensions.py` file to just add a simple project-local Jinja filter ... so I want to reduce as much required boilerplate as possible here. :)

Some examples from Eleventy:
- https://github.com/seancdavis/seancdavis-com/blob/main/.eleventy.js#L51
- https://gitlab.com/juanfernandes/kuk-sool-won-woodbridge/-/blob/master/.eleventy.js
- https://github.com/11ty/eleventy-navigation/blob/master/.eleventy.js (how an installable plugin is written)

I like how the interaction model is *really clear* with plugins here. (Also, worth noting I do *not* like how the actual website and routing configuration is done in JS – I definitely want to keep that in our YAML files.) But the plugins -- since that involves loading JS code that's either written locally or installed via NPM, I like how that is just non-magical and straightforward.

We also have something similar that @stevenle has done in Oak (private repo, FYI: Oak overlaps a lot with Amagaki and the Grow Editor but is specifically designed around blog posts at the moment for one specific client):
- https://github.com/grow/oak/blob/main/example/oak.config.js

That said, I really like how this PR introduces the plugin trigger/hook concept and I think that could be a good thing to follow. Maybe we could combine our approaches somehow? A simple `amagaki.js` file where plugins can be added without needing to fuss with YAML, and keeping plugin management in a familiar JS land for users, but with plugins being executed amongst the codebase with triggers...?

Reply:
I'll think over this more and give a more in-depth discussion later, but one of the things I *do* like about having it in the yaml is that the configuration options that control the build are almost all in the one file instead of having two files that you need to check to see how the pod is being configured.

My initial thought is go either way, but not both. Either it is a `js` file that runs *all* of the configuration or we use the `yaml` file that controls *all* of the configuration.

Reply:
Yeah.... that's a good point. I guess there's nothing stopping us from using static data structures in the JS file. My main beef with how other projects do this (and Gulp, Grunt, Webpack, etc.) is the sheer amount of code that needs to be written to just do configuration.

I suppose we could always encourage the best of both worlds where we do have a JS file but within the JS file, the routes, etc. are done like this:

### 1

```
module.exports = function(amagaki) {
  amagaki.setMetadata({
    title: 'My site title',
    foo: 'Foo bar',
  });
  
  amagaki.setLocalization({
    defaultLocale: 'en',
    locales: ['en', 'fr', 'it']
  });
  
  amagaki.setRoutes(...);
  
  amagaki.addPlugin(...);
  
  // etc.
};
```

### 2

Or more simply...

```
module.exports = function(amagaki) {
  amagaki.setConfig({
    metadata: {
      title: 'My site title',
      foo: 'Foo bar',
    },
    localization: {
      defaultLocale: 'en',
      locales: ['en', 'fr', 'it']
    },
    routes: [...]
  });
  amagaki.addPlugin(...);
};
```
### 3

Or even...

```
module.exports.config = {
  metadata: {
    title: 'My site title',
    foo: 'Foo bar',
  },
  localization: {
    defaultLocale: 'en',
    locales: ['en', 'fr', 'it']
  },
  routes: [...]
};

module.exports = function(amagaki) {
  amagaki.addPlugin(...);
};
```

Reply:
Did we decide to go with the Option 3?

I think that looks pretty clean for having the config as data structure and the main export be a method to run on the pod creation that handles adding the plugins. It should also remove the need for the amagaki config to have to find the modules.

Reply:
Yup I think that's a good choice. :) Prior to implementation we should also determine how to define:

1. The structure of the config object and enforcing it. I never liked how in Grow the YAML file was free-form, and if the user entered bad data Grow would just attempt to read it and throw an unspecified error. We should probably validate the config format somehow now.
2. Conditional configs. In Grow it was possible to use `@env` tagging in the podspec. I think that should be possible too in some form. If the config object is returned from a function call, we can let users dynamically construct the config, e.g.:

```
const config = {};
if (pod.env.name === 'prod') {
  config.locales = ['de', 'fr', 'it'];
} 
...
return config;
```

It's not quite as nice as env tagging but it works. It's just a quick initial thought so any other thinking around how to do this is welcome.

Reply:
Dropping this here so we can refer to it later... seems like a good way to do validation of JSON objects is to use jsonschema validation. tsconfig uses it. Some examples:

- https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/tsconfig.json
- https://www.npmjs.com/package/jsonschema
- http://json-schema.org/

Reply:
Have updated the configuration to come from the `amagaki.js` instead of `amagaki.yaml`. This allows for more robust plugin configuration and provides a 'one-stop' place for the configuration of the pod settings and meta data. It happens after the pod env variable is set, so the config file can also do different configuration based on the env.